sdk/python/kfp/compiler/_data_passing_rewriter.py

```diff
@@ -381,3 +383,337 @@ def deconstruct_single_placeholder(s: str) -> List[str]:
     if not re.fullmatch('{{[-._a-zA-Z0-9]+}}', s):
         return None
     return s.lstrip('{').rstrip('}').split('.')
+
+
+def _replace_output_dir_placeholder(command_line: str,
+    output_directory: Optional[str] = None) -> str:
+    """Replaces the output directory placeholder."""
+    if _components.PIPELINE_ROOT_PLACEHOLDER in command_line:
+        if not output_directory:
+            raise ValueError('output_directory of a pipeline must be specified '
+                             'when URI placeholder is used.')
+        return command_line.replace(
+            _components.PIPELINE_ROOT_PLACEHOLDER, output_directory)
+    return command_line
+
+
+def _refactor_outputs_if_uri_placeholder(
+    container_template: Dict[str, Any],
+    output_to_filename: Dict[str, str]
+) -> None:
+    """Rewrites the output of the container in case of URI placeholder.
+
+    Also, collects the mapping from the output names to output file name.
+
+    Args:
+        container_template: Container template structure.
```

Comment:
nit: "Container template structure." -> "The container template structure." ? 
and same for the below occurrences.